distributed/scheduler.py

```diff
@@ -2066,6 +2070,40 @@ def transition_no_worker_memory(
                 pdb.set_trace()
             raise
 
+    def cogroup_objective(self, cogroup: int, ws: WorkerState) -> tuple:
+        # Cogroups are not always connected subgraphs but if we assume they
+        # were, only the top prio task would need a transfer
+        tasks_in_group = self.cogroups[cogroup]
+        # TODO: this could be made more efficient / we should remeber max if it is required
+        ts_top_prio = max(tasks_in_group, key=lambda ts: ts.priority)
+        dts: TaskState
+        comm_bytes: int = 0
+        cotasks_on_worker = 0
+        for ts in tasks_in_group:
+            if ts in ws.processing or ws in ts.who_has:
+                cotasks_on_worker += 1
+        for dts in ts_top_prio.dependencies:
+            if (
+                # This is new compared to worker_objective
+                (dts not in tasks_in_group or dts not in ws.processing)
+                and ws not in dts.who_has
+            ):
+                nbytes = dts.get_nbytes()
+                comm_bytes += nbytes
+
+        stack_time: float = ws.occupancy / ws.nthreads
+        start_time: float = stack_time + comm_bytes / self.bandwidth
+
+        if ts_top_prio.actor:
+            raise NotImplementedError("Cogroup assignment for actors not implemented")
+        else:
+            return (-cotasks_on_worker, start_time, ws.nbytes)
```

Comment:
So this says, prefer to pick a worker which already has many tasks in the same group even if the start time and required comms are very large.

Reply:
haven't put much thought into this but basically whenever there are tasks in the same cogroup assigned to a worker, this worker will be preferred and otherwise we use the ordinary logic.

Please consider the entire `decide_worker_cogroup + cogroup_objective` block as highly experimental. This PR is more about the cogroup algorithm itself. This is just a cheap way of integrating it in the existing scheduling